File: Pythoncode/300.最长上升子序列.py
class Solution:
    def lengthOfLIS(self, nums):
        if len(nums) == 0:
            return 0
        dp = {} # dp[i] = 以nums[i]作结尾的最长上升子序列的长度
        for i in range(len(nums)):
            if i == 0:
                dp[0] = 1
            else:
                dp[i] = max(dp[j] if nums[j] < nums[i] else 0 for j in range(i)) + 1
        return max(dp.values())

s = Solution()
s.lengthOfLIS([10,9,2,5,3,7,101,18])

File: Cppcode/300.最长上升子序列.cpp
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
    //https://leetcode-cn.com/problems/longest-increasing-subsequence/solution/dong-tai-gui-hua-er-fen-cha-zhao-tan-xin-suan-fa-p/
public:
    int lengthOfLIS(vector<int>& nums) {
        if (nums.empty()) return 0;
        vector<int> dp;
        for (auto i = 0; i < nums.size(); i++)
        {
            auto maxlength = 1;
            for (auto j = 0; j < i; j++)
            {
                if (nums[j] < nums[i])
                {
                    maxlength = max(maxlength, dp[j]+1);
                }   
            }
            dp.push_back(maxlength);
        }
        return *max_element(dp.begin(), dp.end());
    }
};